Helm template throws stack overflow error

**Describe the bug**
We have the following example values.yaml file for helm chart:
```
dashboards:
 clustersmonitoringjson:
   {

         "select": [
            [
              {
                "params": [
                  "status"
                ],
                "type": "field"
              }
            ]
         ]
    }
```
If I want to pass them through `helm template  --set` command stack overflow error is thrown.

Helm command:
```
helm template . --namespace default --set dashboards.clustersmonitoringjson.select[0][0].params[0]=status --set dashboards.clustersmonitoringjson.select[0][0].type=field --name helmguestbook
```
Error thrown:

```
runtime: goroutine stack exceeds 1000000000-byte limit\nfatal error: stack overflow\n\nruntime stack:\nruntime.throw(0x16a2c72, 0xe)\n\t/usr/local/go/src/runtime/panic.go:608 +0x72\nruntime.newstack()\n\t/usr/local/go/src/runtime/stack.go:1008 +0x729\nruntime.morestack()\n\t/usr/local/go/src/runtime/asm_amd64.s:429 +0x8f\n\ngoroutine 1 [running]:\nruntime.efaceeq(0x1660160, 0x141ad00, 0x141ad00, 0x0)\n\t/usr/local/go/src/runtime/alg.go:218 +0x131 fp=0xc0209b0358 sp=0xc0209b0350 pc=0x403631\nruntime.nilinterequal(0xc0209b0448, 0xc00047e848, 0xc5b27c6c1b7408fc)\n\t/usr/local/go/src/runtime/alg.go:216 +0x5d fp=0xc0209b0388 sp=0xc0209b0358 pc=0x4034ed\nruntime.mapaccess2(0x14a80a0, 0xc000509530, 0xc0209b0448, 0x0, 0x0)\n\t/usr/local/go/src/runtime/map.go:479 +0x146 fp=0xc0209b03d0 sp=0xc0209b0388 pc=0x40d646\nsync.(*Map).Load(0x2729c00, 0x1660160, 0x141ad
```

It can be tested with any chart. This error does not depend on the chart itself.
The issue is not reproducible if parameters are separated by commas in one --set parameters.

I was able to reproduce the issue using the following command for Helm 3.

```
><> helm template stable/wordpress --set dashboards.clustersmonitoringjson.select[0][0].params[0]=status --set dashboards.clustersmonitoringjson.select[0][0].type=field
panic: interface conversion: interface {} is []interface {}, not map[string]interface {}

goroutine 1 [running]:
helm.sh/helm/v3/pkg/strvals.(*parser).listItem(0xc0062928a0, 0xc0003ce8f0, 0x1, 0x1, 0x0, 0x0, 0xc00000005b, 0x0, 0x0, 0x0)
        /home/circleci/helm.sh/helm/pkg/strvals/parser.go:288 +0xdbf
helm.sh/helm/v3/pkg/strvals.(*parser).listItem(0xc0062928a0, 0xc0003ce8f0, 0x1, 0x1, 0x0, 0x1, 0x5b, 0x0, 0x0, 0x203000)
        /home/circleci/helm.sh/helm/pkg/strvals/parser.go:282 +0x3f1
helm.sh/helm/v3/pkg/strvals.(*parser).key(0xc0062928a0, 0xc000580960, 0xc0036c56a8, 0x16)
        /home/circleci/helm.sh/helm/pkg/strvals/parser.go:174 +0xb4d
helm.sh/helm/v3/pkg/strvals.(*parser).key(0xc0062928a0, 0xc0005808d0, 0xc0036c5848, 0xa)
        /home/circleci/helm.sh/helm/pkg/strvals/parser.go:212 +0x1ad
helm.sh/helm/v3/pkg/strvals.(*parser).key(0xc0062928a0, 0xc000580840, 0x152e35d, 0x17c9880)
        /home/circleci/helm.sh/helm/pkg/strvals/parser.go:212 +0x1ad
helm.sh/helm/v3/pkg/strvals.(*parser).parse(0xc0062928a0, 0xc000580840, 0x0)
        /home/circleci/helm.sh/helm/pkg/strvals/parser.go:130 +0x38
helm.sh/helm/v3/pkg/strvals.ParseInto(0x7ffe61d0c376, 0x39, 0xc000580840, 0x0, 0x0)
        /home/circleci/helm.sh/helm/pkg/strvals/parser.go:70 +0xbe
helm.sh/helm/v3/pkg/cli/values.(*Options).MergeValues(0xc0000de420, 0xc0062925a0, 0x1, 0x1, 0x0, 0x0, 0x0)
        /home/circleci/helm.sh/helm/pkg/cli/values/options.go:62 +0x22b
main.runInstall(0xc0003c8960, 0x1, 0x5, 0xc00008f9e0, 0xc0000de420, 0x1b7c9e0, 0xc0000dc000, 0xc0003f8dc8, 0x31, 0x30)
        /home/circleci/helm.sh/helm/cmd/helm/install.go:168 +0x331
main.newTemplateCmd.func1(0xc00027d900, 0xc0003c8960, 0x1, 0x5, 0x0, 0x0)
        /home/circleci/helm.sh/helm/cmd/helm/template.go:63 +0x117
github.com/spf13/cobra.(*Command).execute(0xc00027d900, 0xc0003c8910, 0x5, 0x5, 0xc00027d900, 0xc0003c8910)
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002b9900, 0x1bba9e0, 0xc00020e160, 0x191d249)
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
        /home/circleci/helm.sh/helm/cmd/helm/helm.go:75 +0x204
```

--- a/pkg/strvals/parser.go
+++ b/pkg/strvals/parser.go
@@ -305,8 +305,15 @@ func (t *parser) listItem(list []interface{}, i int) ([]interface{}, error) {
 		if err != nil {
 			return list, errors.Wrap(err, "error parsing index")
 		}
+		var crtList []interface{}
+		if len(list) > i {
+			// If nested list already exists, take the value of list to next cycle.
+			crtList = list[i].([]interface{})
+		} else {
+			crtList = list
+		}
 		// Now we need to get the value after the ].
-		list2, err := t.listItem(list, i)
+		list2, err := t.listItem(crtList, i)
 		if err != nil {
 			return list, err
 		}

--- a/pkg/strvals/parser.go
+++ b/pkg/strvals/parser.go
@@ -314,16 +314,11 @@ func (t *parser) listItem(list []interface{}, i int) ([]interface{}, error) {
 			}
 		}
 		// Now we need to get the value after the ].
-<<<<<<< HEAD
-		list2, err := t.listItem(crtList, i)
+		list2, err := t.listItem(crtList, nextI)
 		if err != nil {
 			return list, err
 		}
 		return setIndex(list, i, list2)
-=======
-		list2, err := t.listItem(crtList, nextI)
-		return setIndex(list, i, list2), err
->>>>>>> fix another extreme case
 	case last == '.':
 		// We have a nested object. Send to t.key
 		inner := map[string]interface{}{}

--- a/pkg/strvals/parser.go
+++ b/pkg/strvals/parser.go
@@ -301,7 +301,7 @@ func (t *parser) listItem(list []interface{}, i int) ([]interface{}, error) {
 		}
 	case last == '[':
 		// now we have a nested list. Read the index and handle.
-		i, err := t.keyIndex()
+		nextI, err := t.keyIndex()
 		if err != nil {
 			return list, errors.Wrap(err, "error parsing index")
 		}
@@ -309,15 +309,18 @@ func (t *parser) listItem(list []interface{}, i int) ([]interface{}, error) {
 		if len(list) > i {
 			// If nested list already exists, take the value of list to next cycle.
 			crtList = list[i].([]interface{})
-		} else {
-			crtList = list
 		}
 		// Now we need to get the value after the ].
+<<<<<<< HEAD
 		list2, err := t.listItem(crtList, i)
 		if err != nil {
 			return list, err
 		}
 		return setIndex(list, i, list2)
+=======
+		list2, err := t.listItem(crtList, nextI)
+		return setIndex(list, i, list2), err
+>>>>>>> fix another extreme case
 	case last == '.':
 		// We have a nested object. Send to t.key
 		inner := map[string]interface{}{}

--- a/pkg/strvals/parser.go
+++ b/pkg/strvals/parser.go
@@ -308,7 +308,10 @@ func (t *parser) listItem(list []interface{}, i int) ([]interface{}, error) {
 		var crtList []interface{}
 		if len(list) > i {
 			// If nested list already exists, take the value of list to next cycle.
-			crtList = list[i].([]interface{})
+			existed := list[i]
+			if existed != nil {
+				crtList = list[i].([]interface{})
+			}
 		}
 		// Now we need to get the value after the ].
 <<<<<<< HEAD

--- a/pkg/strvals/parser_test.go
+++ b/pkg/strvals/parser_test.go
@@ -427,7 +427,7 @@ func TestParseInto(t *testing.T) {
 			"inner2": "value2",
 		},
 	}
-	input := "outer.inner1=value1,outer.inner3=value3,outer.inner4=4"
+	input := "outer.inner1=value1,outer.inner3=value3,outer.inner4=4,listOuter[0][0].type=listValue"
 	expect := map[string]interface{}{
 		"outer": map[string]interface{}{
 			"inner1": "value1",
@@ -435,9 +435,19 @@ func TestParseInto(t *testing.T) {
 			"inner3": "value3",
 			"inner4": 4,
 		},
+		"listOuter": [][]interface{}{{map[string]string{
+			"type":   "listValue",
+			"status": "alive",
+		}},
+		},
 	}
 
 	if err := ParseInto(input, got); err != nil {
+		t.Fatal(err)
+	}
+
+	input2 := "listOuter[0][0].status=alive"
+	if err := ParseInto(input2, got); err != nil {
 		t.Fatal(err)
 	}
 

--- a/pkg/strvals/parser_test.go
+++ b/pkg/strvals/parser_test.go
@@ -421,49 +421,118 @@ func TestParseSet(t *testing.T) {
 }
 
 func TestParseInto(t *testing.T) {
-	got := map[string]interface{}{
-		"outer": map[string]interface{}{
-			"inner1": "overwrite",
-			"inner2": "value2",
+	tests := []struct {
+		input  string
+		input2 string
+		got    map[string]interface{}
+		expect map[string]interface{}
+		err    bool
+	}{
+		{
+			input: "outer.inner1=value1,outer.inner3=value3,outer.inner4=4",
+			got: map[string]interface{}{
+				"outer": map[string]interface{}{
+					"inner1": "overwrite",
+					"inner2": "value2",
+				},
+			},
+			expect: map[string]interface{}{
+				"outer": map[string]interface{}{
+					"inner1": "value1",
+					"inner2": "value2",
+					"inner3": "value3",
+					"inner4": 4,
+				}},
+			err: false,
+		},
+		{
+			input:  "listOuter[0][0].type=listValue",
+			input2: "listOuter[0][0].status=alive",
+			got:    map[string]interface{}{},
+			expect: map[string]interface{}{
+				"listOuter": [][]interface{}{{map[string]string{
+					"type":   "listValue",
+					"status": "alive",
+				}}},
+			},
+			err: false,
+		},
+		{
+			input:  "listOuter[0][0].type=listValue",
+			input2: "listOuter[1][0].status=alive",
+			got:    map[string]interface{}{},
+			expect: map[string]interface{}{
+				"listOuter": [][]interface{}{
+					{
+						map[string]string{"type": "listValue"},
+					},
+					{
+						map[string]string{"status": "alive"},
+					},
+				},
+			},
+			err: false,
+		},
+		{
+			input:  "listOuter[0][1][0].type=listValue",
+			input2: "listOuter[0][0][1].status=alive",
+			got: map[string]interface{}{
+				"listOuter": []interface{}{
+					[]interface{}{
+						[]interface{}{
+							map[string]string{"exited": "old"},
+						},
+					},
+				},
+			},
+			expect: map[string]interface{}{
+				"listOuter": [][][]interface{}{
+					{
+						{
+							map[string]string{"exited": "old"},
+							map[string]string{"status": "alive"},
+						},
+						{
+							map[string]string{"type": "listValue"},
+						},
+					},
+				},
+			},
+			err: false,
 		},
 	}
-	input := "outer.inner1=value1,outer.inner3=value3,outer.inner4=4,listOuter[0][0].type=listValue"
-	expect := map[string]interface{}{
-		"outer": map[string]interface{}{
-			"inner1": "value1",
-			"inner2": "value2",
-			"inner3": "value3",
-			"inner4": 4,
-		},
-		"listOuter": [][]interface{}{{map[string]string{
-			"type":   "listValue",
-			"status": "alive",
-		}},
-		},
-	}
+	for _, tt := range tests {
+		if err := ParseInto(tt.input, tt.got); err != nil {
+			t.Fatal(err)
+		}
+		if tt.err {
+			t.Errorf("%s: Expected error. Got nil", tt.input)
+		}
 
-	if err := ParseInto(input, got); err != nil {
-		t.Fatal(err)
-	}
+		if tt.input2 != "" {
+			if err := ParseInto(tt.input2, tt.got); err != nil {
+				t.Fatal(err)
+			}
+			if tt.err {
+				t.Errorf("%s: Expected error. Got nil", tt.input2)
+			}
+		}
 
-	input2 := "listOuter[0][0].status=alive"
-	if err := ParseInto(input2, got); err != nil {
-		t.Fatal(err)
-	}
+		y1, err := yaml.Marshal(tt.expect)
+		if err != nil {
+			t.Fatal(err)
+		}
+		y2, err := yaml.Marshal(tt.got)
+		if err != nil {
+			t.Fatalf("Error serializing parsed value: %s", err)
+		}
 
-	y1, err := yaml.Marshal(expect)
-	if err != nil {
-		t.Fatal(err)
-	}
-	y2, err := yaml.Marshal(got)
-	if err != nil {
-		t.Fatalf("Error serializing parsed value: %s", err)
-	}
-
-	if string(y1) != string(y2) {
-		t.Errorf("%s: Expected:\n%s\nGot:\n%s", input, y1, y2)
+		if string(y1) != string(y2) {
+			t.Errorf("%s: Expected:\n%s\nGot:\n%s", tt.input, y1, y2)
+		}
 	}
 }
+
 func TestParseIntoString(t *testing.T) {
 	got := map[string]interface{}{
 		"outer": map[string]interface{}{

